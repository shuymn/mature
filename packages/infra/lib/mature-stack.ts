import * as path from "path";
import * as events from "@aws-cdk/aws-events";
import * as targets from "@aws-cdk/aws-events-targets";
import * as iam from "@aws-cdk/aws-iam";
import * as lambda from "@aws-cdk/aws-lambda";
import * as cdk from "@aws-cdk/core";

export class MatureStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const matureExecutionRole = new iam.Role(this, "mature-execution-role", {
      roleName: "mature-execution-role",
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole"
        ),
      ],
    });

    const stack = cdk.Stack.of(this);
    matureExecutionRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["ssm:GetParameters"],
        resources: [
          `arn:aws:ssm:${stack.region}:${stack.account}:parameter/mature/production/*`,
        ],
      })
    );

    matureExecutionRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["cloudwatch:PutMetricData"],
        resources: ["*"],
        conditions: {
          StringEquals: {
            "cloudwatch:namespace": "NatureRemo/RoomMetrics",
          },
        },
      })
    );

    const matureFn = new lambda.Function(this, "mature-function", {
      functionName: "mature",
      code: lambda.Code.fromAsset(path.resolve(__dirname, "../../mature/dist")),
      handler: "mature",
      runtime: lambda.Runtime.GO_1_X,
      memorySize: 256,
      timeout: cdk.Duration.seconds(30),
      role: matureExecutionRole,
      tracing: lambda.Tracing.ACTIVE,
      currentVersionOptions: {
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retryAttempts: 0,
      },
      environment: {
        MATURE_ACCESS_TOKEN_KEY: "/mature/production/access-token",
        MATURE_DEVICE_ID_KEY: "/mature/production/device-id/main-room",
      },
    });

    matureFn.currentVersion.addAlias("development");

    const matureFnProdVersion = lambda.Version.fromVersionArn(
      this,
      "mature-function-version-production",
      `${matureFn.functionArn}:7`
    );
    const matureFnProdAlias = matureFnProdVersion.addAlias("production");

    const matureRule = new events.Rule(this, "mature-rule", {
      ruleName: "mature-rule",
      schedule: events.Schedule.expression("rate(1 minute)"),
    });

    matureRule.addTarget(new targets.LambdaFunction(matureFnProdAlias));
  }
}
