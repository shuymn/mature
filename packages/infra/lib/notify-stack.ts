import * as path from "path";
import * as events from "@aws-cdk/aws-events";
import * as targets from "@aws-cdk/aws-events-targets";
import * as iam from "@aws-cdk/aws-iam";
import * as lambda from "@aws-cdk/aws-lambda";
import * as cdk from "@aws-cdk/core";

export class NotifyStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const role = new iam.Role(this, "notify-execution-role", {
      roleName: "mature-notify-execution-role",
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole"
        ),
      ],
    });

    const stack = cdk.Stack.of(this);
    role.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["ssm:GetParameters"],
        resources: [
          `arn:aws:ssm:${stack.region}:${stack.account}:parameter/mature/production/slack-token`,
          `arn:aws:ssm:${stack.region}:${stack.account}:parameter/mature/production/channel-id`,
        ],
      })
    );

    role.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["cloudwatch:GetMetricWidgetImage"],
        resources: ["*"],
      })
    );

    const fn = new lambda.Function(this, "notify-function", {
      functionName: "mature-notify",
      code: lambda.Code.fromAsset(path.resolve(__dirname, "../../notify/dist")),
      handler: "notify",
      runtime: lambda.Runtime.GO_1_X,
      memorySize: 128,
      timeout: cdk.Duration.seconds(30),
      role: role,
      tracing: lambda.Tracing.ACTIVE,
      currentVersionOptions: {
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retryAttempts: 0,
      },
      environment: {
        MATURE_SLACK_TOKEN_KEY: "/mature/production/slack-token",
        MATURE_SLACK_CHANNEL_ID_KEY: "/mature/production/channel-id",
      },
    });

    fn.currentVersion.addAlias("development");

    const prodVersion = lambda.Version.fromVersionArn(
      this,
      "notify-function-version-production",
      `${fn.functionArn}:${fn.currentVersion.version}`
    );
    const prodAlias = prodVersion.addAlias("production");

    const rule = new events.Rule(this, "notify-rule", {
      ruleName: "mature-notify-rule",
      schedule: events.Schedule.expression("cron(0 15 * * ? *)"),
    });

    rule.addTarget(new targets.LambdaFunction(prodAlias));
  }
}
