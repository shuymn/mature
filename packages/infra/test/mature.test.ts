import "@aws-cdk/assert/jest";
import * as cdk from "@aws-cdk/core";
import { MatureStack } from "../lib/mature-stack";

describe("fine-grained assertions", () => {
  test("iam role", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Version: "2012-10-17",
        Statement: [
          {
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          },
        ],
      },
      ManagedPolicyArns: [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                Ref: "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            ],
          ],
        },
      ],
      RoleName: "mature-execution-role",
    });
  });

  test("iam policy", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::IAM::Policy", {
      PolicyDocument: {
        Statement: [
          {
            Action: "ssm:GetParameters",
            Effect: "Allow",
            Resource: [
              {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      Ref: "AWS::Region",
                    },
                    ":",
                    {
                      Ref: "AWS::AccountId",
                    },
                    ":parameter/mature/production/access-token",
                  ],
                ],
              },
              {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      Ref: "AWS::Region",
                    },
                    ":",
                    {
                      Ref: "AWS::AccountId",
                    },
                    ":parameter/mature/production/device-id/*",
                  ],
                ],
              },
            ],
          },
          {
            Action: "cloudwatch:PutMetricData",
            Condition: {
              StringEquals: {
                "cloudwatch:namespace": "NatureRemo/RoomMetrics",
              },
            },
            Effect: "Allow",
            Resource: "*",
          },
          {
            Action: ["xray:PutTraceSegments", "xray:PutTelemetryRecords"],
            Effect: "Allow",
            Resource: "*",
          },
        ],
        Version: "2012-10-17",
      },
    });
  });

  test("lambda function", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::Lambda::Function", {
      Handler: "mature",
      Runtime: "go1.x",
      Environment: {
        Variables: {
          MATURE_ACCESS_TOKEN_KEY: "/mature/production/access-token",
          MATURE_DEVICE_ID_KEY: "/mature/production/device-id/main-room",
        },
      },
      FunctionName: "mature",
      MemorySize: 256,
      Timeout: 30,
      TracingConfig: {
        Mode: "Active",
      },
    });
  });

  test("lambda event invoke config", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::Lambda::EventInvokeConfig", {
      MaximumRetryAttempts: 0,
    });
  });

  test("lambda alias", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::Lambda::Alias", {
      Name: "production",
    });
  });

  test("event rule", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::Events::Rule", {
      Name: "mature-rule",
      ScheduleExpression: "rate(1 minute)",
      State: "ENABLED",
    });
  });

  test("lambda permission", () => {
    const app = new cdk.App();
    const stack = new MatureStack(app, "TestMatureStack");

    expect(stack).toHaveResource("AWS::Lambda::Permission", {
      Action: "lambda:InvokeFunction",
      Principal: "events.amazonaws.com",
    });
  });
});
