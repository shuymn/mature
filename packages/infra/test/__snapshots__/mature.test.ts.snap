// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test match 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791ArtifactHashA4A1C80E": Object {
      "Description": "Artifact hash for asset \\"c0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791\\"",
      "Type": "String",
    },
    "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791S3BucketE32087DA": Object {
      "Description": "S3 bucket for asset \\"c0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791\\"",
      "Type": "String",
    },
    "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791S3VersionKey24FAAAD6": Object {
      "Description": "S3 key for asset version \\"c0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "matureexecutionrole7C1F1FE4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "mature-execution-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "matureexecutionroleDefaultPolicyB02C4895": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameters",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:ssm:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":parameter/mature/production/access-token",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:ssm:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":parameter/mature/production/device-id/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Condition": Object {
                "StringEquals": Object {
                  "cloudwatch:namespace": "NatureRemo/RoomMetrics",
                },
              },
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "matureexecutionroleDefaultPolicyB02C4895",
        "Roles": Array [
          Object {
            "Ref": "matureexecutionrole7C1F1FE4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "maturefunctionCurrentVersion7961BCFE4d6f5ceae02142652d931542c4361806": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "maturefunctionD957EC05",
        },
      },
      "Type": "AWS::Lambda::Version",
      "UpdateReplacePolicy": "Retain",
    },
    "maturefunctionCurrentVersionAliasdevelopment0024E1F3": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "maturefunctionD957EC05",
        },
        "FunctionVersion": Object {
          "Fn::GetAtt": Array [
            "maturefunctionCurrentVersion7961BCFE4d6f5ceae02142652d931542c4361806",
            "Version",
          ],
        },
        "Name": "development",
      },
      "Type": "AWS::Lambda::Alias",
    },
    "maturefunctionCurrentVersionEventInvokeConfigD3994A1F": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "maturefunctionD957EC05",
        },
        "MaximumRetryAttempts": 0,
        "Qualifier": Object {
          "Fn::GetAtt": Array [
            "maturefunctionCurrentVersion7961BCFE4d6f5ceae02142652d931542c4361806",
            "Version",
          ],
        },
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "maturefunctionD957EC05": Object {
      "DependsOn": Array [
        "matureexecutionroleDefaultPolicyB02C4895",
        "matureexecutionrole7C1F1FE4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791S3BucketE32087DA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791S3VersionKey24FAAAD6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc0e183d4d033462ab23af64f704ef92eb2a163d7f9f4ef841443adb4f6225791S3VersionKey24FAAAD6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "MATURE_ACCESS_TOKEN_KEY": "/mature/production/access-token",
            "MATURE_DEVICE_ID_KEY": "/mature/production/device-id/main-room",
          },
        },
        "FunctionName": "mature",
        "Handler": "mature",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "matureexecutionrole7C1F1FE4",
            "Arn",
          ],
        },
        "Runtime": "go1.x",
        "Timeout": 30,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "maturefunctionversionproductionAliasproduction32297CA3": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Fn::Select": Array [
            6,
            Object {
              "Fn::Split": Array [
                ":",
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "maturefunctionD957EC05",
                          "Arn",
                        ],
                      },
                      ":",
                      Object {
                        "Fn::GetAtt": Array [
                          "maturefunctionCurrentVersion7961BCFE4d6f5ceae02142652d931542c4361806",
                          "Version",
                        ],
                      },
                    ],
                  ],
                },
              ],
            },
          ],
        },
        "FunctionVersion": Object {
          "Fn::Select": Array [
            7,
            Object {
              "Fn::Split": Array [
                ":",
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "maturefunctionD957EC05",
                          "Arn",
                        ],
                      },
                      ":",
                      Object {
                        "Fn::GetAtt": Array [
                          "maturefunctionCurrentVersion7961BCFE4d6f5ceae02142652d931542c4361806",
                          "Version",
                        ],
                      },
                    ],
                  ],
                },
              ],
            },
          ],
        },
        "Name": "production",
      },
      "Type": "AWS::Lambda::Alias",
    },
    "maturerule923D5378": Object {
      "Properties": Object {
        "Name": "mature-rule",
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "maturefunctionversionproductionAliasproduction32297CA3",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "matureruleAllowEventRuleTestMatureStackmaturerule1D2183119212A4EF": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "maturefunctionversionproductionAliasproduction32297CA3",
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "maturerule923D5378",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
